### PR TITLE
[clang-tidy] new check misc-use-internal-linkage

### PR DESCRIPTION
**Pull request by HerrCai0907**
Add new check misc-use-internal-linkage to detect variable and function can be marked as static.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/readability/MarkStaticCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+void MarkStaticCheck::registerMatchers(MatchFinder *Finder) {
+  auto Common =
+      allOf(isFirstDecl(), isInMainFile(),
+            unless(anyOf(isStaticStorageClass(), isExternStorageClass(),
+                         isInAnonymousNamespace())));
+  Finder->addMatcher(
+      functionDecl(Common, unless(cxxMethodDecl()), unless(isMain()))
+          .bind("fn"),
+      this);
+  Finder->addMatcher(varDecl(Common, hasGlobalStorage()).bind("var"), this);
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/MarkStaticCheck.cpp:50]: Few things are missing here, I wrote same check already for private usage, so let me drop it here (for ideas):
```
  Finder->addMatcher(functionDecl(isExpansionInMainFile(),
                                  isDefinition(),
                                  hasExternalFormalLinkage(),
                                  unless(isMain()),
                                  unless(isExternC()),
                                  unless(isExplicitTemplateSpecialization()),
                                  unless(clang::ast_matchers::isTemplateInstantiation()),
                                  unless(cxxMethodDecl()),
                                  unless(hasAncestor(friendDecl())),
                                  unless(hasDeclarationOutsideMainFile())
                                 ).bind("fun"), this);
  Finder->addMatcher(varDecl(isExpansionInMainFile(),
                             isDefinition(),
                             hasGlobalStorage(),
                             unless(isStaticLocal()),
                             unless(isStaticStorageClass()),
                             hasExternalFormalLinkage(),
                             unless(isExternC()),
                             unless(hasDeclarationOutsideMainFile()),
                             unless(hasVarRedeclaration(anyOf(isStaticStorageClass(), isExternC())))
                            ).bind("var"), this);

```

```
AST_MATCHER(Decl, hasDeclarationOutsideMainFile)
{
    auto &SourceManager = Finder->getASTContext().getSourceManager();
    for(const auto* l_it : Node.redecls())
    {
        if (not SourceManager.isInMainFile(SourceManager.getExpansionLoc(l_it->getBeginLoc())))
        {
            return true;
        }
    }

    return not SourceManager.isInMainFile(SourceManager.getExpansionLoc(Node.getBeginLoc()));
}

AST_MATCHER_P(VarDecl, hasVarRedeclaration, ast_matchers::internal::Matcher<VarDecl>, InnerMatcher)
{
    for(const auto* l_it : Node.redecls())
    {
        if (InnerMatcher.matches(*l_it, Finder, Builder))
        {
            return true;
        }
    }

    return false;
}
```

You need also take into account unity builds, something that I didn't took into consideration, when function/variable is in .cpp file, but thats not a a main file. I think there is some API in Clang-tify for that.

But at same time, I were focusing more on external linkage:
`function %0 has external linkage but it's used only in this source file and has no public declaration, make it internal (anonymous namespace or static) or add public declaration in header file"`

File: clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+readability-mark-static
+=======================
+
+Detects variable and function can be marked as static.
+
+Static functions and variables are scoped to a single file. Marking functions
+and variables as static helps to better remove dead code. In addition, it gives
+the compiler more information and can help compiler make more aggressive
+optimizations.
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst:12]: Mention other benefits: avoid ODR violations

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
1. Wrong check name, maybe readability-unessesary-external-linkage or something, maybe it should even be an performance check [as there it will bring more benefits] (current check suggest more an "static methods").
2. Add support for unity builds (.cpp files included from single .cpp file = isInMainFile is not sufficient)
3. Diagnostic need to be more detailed
4. When I run into these issues, i got questions from developers why not use anonymous namespaces (maybe it could be configurable, or finall check restrict only to static [I prefer static])

Additionally when those issues were being fixed in project, there were lot of situations where there were forward declaration in header file, but developer forget to include header file in .cpp, this is why diagnostic need to be more detailed, so developer could find issue, instead of adding static.

In summary, my main concern is check name & diagnostic, current name is too generic

**Comment by HerrCai0907** [comment]
> Add support for unity builds (.cpp files included from single .cpp file = isInMainFile is not sufficient)

For unity build, it looks like a project level decision. Then they can ignore this check. Maybe I can provide an option to treat some extension as main file also.

**Comment by firewave** [comment]
Isn't this already covered by `-Wmissing-prototypes`?

**Comment by PiotrZSL** [comment]
> Isn't this already covered by `-Wmissing-prototypes`?

No, because -Wmissing-prototypes will not provide warning, when both declaration and definition is in .cpp

**Comment by PiotrZSL** [comment]
> Maybe I can provide an option to treat some extension as main file also.

Clang-tidy already got those options, check just need to utilize them.

File: clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst (hunk 2 above)
Comment (PiotrZSL) [reply clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst:12]: Yes it does, because you cannot get access to such variable/function via extern in separate TU.

[(2.3)](https://eel.is/c++draft/basic.link#2.3)
When a name has [internal linkage](https://eel.is/c++draft/basic.link#def:linkage,internal), the entity it denotes can be referred to by names from other scopes in the same translation unit[.](https://eel.is/c++draft/basic.link#2.3.sentence-1)

The name of an entity that belongs to a [namespace scope](https://eel.is/c++draft/basic.scope.namespace) has internal linkage if it is the name of
[(3.1)](https://eel.is/c++draft/basic.link#3.1)
a variable, variable template, function, or function template that is explicitly declared static; or

Comment (firewave) [reply clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst:12]: Did some research but did not find the case I remembered (it might have variable templates though):
- anonymous namespaces do not have external linkage until C++11: https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces / https://en.cppreference.com/w/cpp/language/storage_duration#Linkage
- `static` variable templates do not have internal linkage until C++14: https://en.cppreference.com/w/cpp/language/storage_duration#Linkage

Interesting tidbit that enumerations have external linkage.

Comment (PiotrZSL) [reply clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst:12]: Any of the following names declared at namespace scope have internal linkage:
- variables, variable templates(since C++14), functions, or function templates declared static;

Looks fine... Still would be nice to allow user select static or anonymous namespace for "auto fix"

**Comment by EugeneZelenko** [comment]
`readability-use-static`?

File: clang-tools-extra/clang-tidy/readability/UnnecessaryExternalLinkageCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                                                        VarDecl)) {
+  return Node.isFirstDecl();
+}
+
+AST_MATCHER(Decl, isInMainFile) {
+  for (const Decl *D : Node.redecls())
+    if (!Finder->getASTContext().getSourceManager().isInMainFile(
+            D->getLocation()))
+      return false;
+  return true;
```
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/readability/UnnecessaryExternalLinkageCheck.cpp:33]: Can be replaced by `std::all_of(...)`.

File: clang-tools-extra/clang-tidy/readability/UnnecessaryExternalLinkageCheck.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      functionDecl(Common, unless(cxxMethodDecl()), unless(isMain()))
+          .bind("fn"),
+      this);
+  Finder->addMatcher(varDecl(Common, hasGlobalStorage()).bind("var"), this);
+}
+
+static constexpr StringRef Message =
+    "%0 %1 can be internal linkage, "
+    "marking as static or using anonymous namespace can avoid external "
+    "linkage.";
```
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/readability/UnnecessaryExternalLinkageCheck.cpp:66]: A bit shorter:
```suggestion
    "%0 %1 can be made static or moved into an anonymous "
    "namespace to enforce internal linkage";
```

File: clang-tools-extra/clang-tidy/readability/UnnecessaryExternalLinkageCheck.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+void UnnecessaryExternalLinkageCheck::check(const MatchFinder::MatchResult &Result) {
+  if (const auto *FD = Result.Nodes.getNodeAs<FunctionDecl>("fn")) {
+    diag(FD->getLocation(), Message) << "function" << FD;
+    return;
+  }
+  if (const auto *VD = Result.Nodes.getNodeAs<VarDecl>("var")) {
+    diag(VD->getLocation(), Message) << "variable" << VD;
+    return;
+  }
+  llvm_unreachable("");
```
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/readability/UnnecessaryExternalLinkageCheck.cpp:77]: Unnecessary I think.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 6]
```diff
@@ -151,6 +151,11 @@ New checks
   Enforces consistent style for enumerators' initialization, covering three
   styles: none, first only, or all initialized explicitly.
 
+- New :doc:`readability-unnecessary-external-linkage
+  <clang-tidy/checks/readability/unnecessary-external-linkage>` check.
+
+  Detects variable and function can be marked as static.
```
Comment (SimplyDanny) [inline clang-tools-extra/docs/ReleaseNotes.rst:157]:
```suggestion
  Detects variables and functions that can be marked as static
  or moved into an anonymous namespace to enforce internal linkage.
```

File: clang-tools-extra/docs/clang-tidy/checks/readability/unnecessary-external-linkage.rst [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+readability-unnecessary-external-linkage
+========================================
+
+Detects variable and function can be marked as static.
+
+Static functions and variables are scoped to a single file. Marking functions
+and variables as static helps to better remove dead code. In addition, it gives
+the compiler more information and can help compiler make more aggressive
+optimizations.
```
Comment (SimplyDanny) [inline clang-tools-extra/docs/clang-tidy/checks/readability/unnecessary-external-linkage.rst:11]:
```suggestion
the compiler more information and allows for more aggressive optimizations.
```

**Comment by HerrCai0907** [comment]
> * The auto-fix should be configurable to choose `static` or anonymous namespace.

Should I implement auto-fix for this check? Maybe some functions / variables will be marked incorrectly and cause link error because the coder just forget to include the header file but this check mark it as internal linkage.
WDYT? @carlosgalvezp @PiotrZSL @EugeneZelenko @SimplyDanny

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp [hunk 8]
```diff
@@ -1,4 +1,4 @@
-//===--- UnnecessaryExternalLinkageCheck.cpp - clang-tidy
+//===--- UseInternalLinkageCheck.cpp - clang-tidy
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:1]: Please merge into single line.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 9]
```diff
@@ -151,8 +151,8 @@ New checks
   Enforces consistent style for enumerators' initialization, covering three
   styles: none, first only, or all initialized explicitly.
 
-- New :doc:`readability-unnecessary-external-linkage
-  <clang-tidy/checks/readability/unnecessary-external-linkage>` check.
+- New :doc:`misc-use-internal-linkage
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:154]: Please keep alphabetical order (by check name) in this section.

File: clang-tools-extra/docs/clang-tidy/checks/misc/use-internal-linkage.rst [hunk 10]
```diff
@@ -1,7 +1,7 @@
-.. title:: clang-tidy - readability-unnecessary-external-linkage
+.. title:: clang-tidy - misc-use-internal-linkage
 
-readability-unnecessary-external-linkage
-========================================
+misc-use-internal-linkage
+=========================
 
 Detects variable and function can be marked as static.
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/misc/use-internal-linkage.rst:6]: Please synchronize with Release Notes after change there.

**Comment by carlosgalvezp** [comment]
> Should I implement auto-fix for this check?

As a first step we can have it without auto-fix and add that as a second step once we figure out a good way to do it.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  Finder->addMatcher(
+      functionDecl(Common, unless(cxxMethodDecl()), unless(isMain()))
+          .bind("fn"),
+      this);
+  Finder->addMatcher(varDecl(Common, hasGlobalStorage()).bind("var"), this);
+}
+
+static constexpr StringRef Message =
+    "%0 %1 can be can be made static or moved into an anonymous namespace "
+    "to enforce internal linkage.";
```
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:65]: Typically, messages don't end with a period:
```suggestion
    "%0 %1 can be made static or moved into an anonymous namespace "
    "to enforce internal linkage";
```

**Review by SimplyDanny — APPROVED** [review]

File: clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst (hunk 2 above)
Comment (EugeneZelenko) [reply clang-tools-extra/docs/clang-tidy/checks/readability/mark-static.rst]: I'm not sure about automatic fixes for anonymous namespaces. For example, in my work project we keep single monolithic anonymous namespace, so function body move would be required.

**Review by 5chmidti — COMMENTED** [review]
Regarding unit build:
`google-global-names-in-headers` is one of the checks that check file extensions: https://github.com/llvm/llvm-project/blob/d33937b6236767137a1ec3393d0933f10eed4ffe/clang-tools-extra/clang-tidy/google/GlobalNamesInHeadersCheck.cpp#L42-L44

Nice that there is a check for this now, it looks like there are a few instances of this issue in LLVM as well.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+void UseInternalLinkageCheck::check(const MatchFinder::MatchResult &Result) {
+  if (const auto *FD = Result.Nodes.getNodeAs<FunctionDecl>("fn")) {
+    diag(FD->getLocation(), Message) << "function" << FD;
+    return;
+  }
+  if (const auto *VD = Result.Nodes.getNodeAs<VarDecl>("var")) {
+    diag(VD->getLocation(), Message) << "variable" << VD;
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:75]: It won't. It does not provide the source range, then it will only mark the begin location instead of the whole source range. The warning would be like this:
```
a.cpp:3:3: warning: function 'f2' can be made static or moved into an anonymous namespace to enforce internal linkage [misc-use-internal-linkage]
    3 | S f2(const S &val) { return val; }
      |   ^
```

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+AST_MATCHER(Decl, isFirstDecl) { return Node.isFirstDecl(); }
+
+AST_MATCHER_P(Decl, isInMainFile, FileExtensionsSet, HeaderFileExtensions) {
+  return llvm::all_of(Node.redecls(), [&](const Decl *D) {
+    SourceManager &SM = Finder->getASTContext().getSourceManager();
+    const SourceLocation L = D->getLocation();
+    return SM.isInMainFile(L) &&
+           !utils::isSpellingLocInHeaderFile(L, SM, HeaderFileExtensions);
+  });
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: should be:

```
#source files or main file if its not a header.
return utils::isSpellingLocInFile(L, SM, SourceFileExtensions) ||
          (SM.isInMainFile(L) && !utils::isSpellingLocInFile(L, SM, HeaderFileExtensions));
```

isSpellingLocInHeaderFile should be renamed into isSpellingLocInFile, that "Header" is not needed in those functions.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+AST_POLYMORPHIC_MATCHER(isExternStorageClass,
+                        AST_POLYMORPHIC_SUPPORTED_TYPES(FunctionDecl,
+                                                        VarDecl)) {
+  return Node.getStorageClass() == SC_Extern;
+}
+
+} // namespace
+
+void UseInternalLinkageCheck::registerMatchers(MatchFinder *Finder) {
+  auto Common = allOf(isFirstDecl(), isInMainFile(HeaderFileExtensions),
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:45]: isFirstDecl wont work if someone do:

```
void foo()
{
}
#include "foo.h"
```

but thats, fine, just would be good if documentation mention that it checks first declaration

**Comment by PiotrZSL** [comment]
> > * The auto-fix should be configurable to choose `static` or anonymous namespace.
> 
> Should I implement auto-fix for this check? Maybe some functions / variables will be marked incorrectly and cause link error because the coder just forget to include the header file but this check mark it as internal linkage. WDYT? @carlosgalvezp @PiotrZSL @EugeneZelenko @SimplyDanny

YES. I used such check on a project that I work for. In short it found over an 1000 issues, manually fixing them was not an option.
I had an quick fix with adding `static`, and that worked fine, in some places code didn't compile so had to fix those by my self, but that was fine.

Add an option for "static / anonymous namespace", and generate fixes/hints accordingly.
You can use optional values, or enums and by default suggest one or other, and in such state you may not need to provide fixes. In other config state just provide fixes, even if that would mean wrapping every function in separate anonymous namespace or adding static. There allways can be other check or some clang-format option to merge multiple namespaces.
Do not reorder functions, and one can use other. Also static is safer as it's does not change scope, with namespace user may run into issues, but still fixes are needed. You can always mention in documentation that fixes are not perfect and manual intervention may be required. Even if check will do 80% of job, thats already huge help.

**Review by PiotrZSL — COMMENTED** [review]
Few nits.

Missing:
- more proper handling for unity files (.cpp included from .cpp)
- nits
- auto-fixes (any) - in worst case you can provide just fixes with static, but attaching them to notes.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp (hunk 14 above)
Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:45]: It still work fine. The reason I add `isFirstDecl` is that the check will check all redecl and add this limitation can avoid O(n^2) complexity.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp (hunk 13 above)
Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: It will match more cases which I cannot understand the code style.
e.g.
```c++
/// a.cpp
void f(); // decl-1

/// b.cpp
#include <a.cpp>
void f() {} // decl-2
```
The current version will ignore this cases because decl-1 is not in main file.
But the suggestion version will catch this cases because both file is source fill extensions.

Actually I don't understand and didn't meet this code style, so I keep conservative to avoid false positive.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: "because decl-1 is not in main file" but thats still source file and thats a point.

Take into account 2 use cases:
1. user want to run check on a header files, in such case header file will be a main file, and check wont work.
2. user want to run check on unity cpp, where there is single .cpp file generated by cmake that include other cpp files and those files include header files.
In both cases user should be able to do that. 

You already got "allOf(isFirstDecl(), isInMainFile(HeaderFileExtensions)," and thats fine. so check should work only if first declaration is in source file (no additional header) or when declaration and definition is in header file, in such case adding static/anonymous namespace is also expected (but only if user explicitly want to run check on headers - by messing up with compile commands json for example).

Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: I got the point. Our main disagreement is about

> when declaration and definition is in header file, in such case adding static/anonymous namespace is also expected

In my opinion, we should not emit any wrong for the definition in header file. Add static in header file is dangerous and cause potential bug.

e.g.
```c++
/// a.hpp
PREFIX int f() {
  static int a = 1;
  return a++;
}
void b();

/// a.cpp
#include "a.hpp"
#include <iostream>

int main() {
  std::cout << __func__ << " " << f() << "\n";
  b();
  std::cout << __func__ << " " << f() << "\n";
  b();
}

/// b.cpp
#include "a.hpp"
#include <iostream>

void b() { std::cout << __func__ << " " << f() << "\n"; }
```

if `PREFIX` is `inline`, result is `1 2 3 4`.
if `PREFIX` is `static`, result is `1 2 1 2`.
if `PREFIX` is `inline static`, result is `1 2 1 2`.

if `PREFIX` is `template <int>`, result is `1 2 3 4`.
if `PREFIX` is `template <int> static`, result is `1 2 1 2`.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: I'm fine for lack of fixes in header files.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: this still won't work for "unity files" (includes .cpp)

Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: what is the expected behavior for unity files? use static or not? I am a little bit confused.

Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: As a summary
If main file include non-header file (*.inc, *.cpp, etc..), this check should treat it as main file.
Then we just need to check there are no `redecl` in header file.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:33]: yes, that FirstDecl stuff will work. just note that we got config for header extensions and source extensions. better would be to use source. as those .inc could be part of header files also.

**Review by PiotrZSL — APPROVED** [review]
I accept this change as it's fine for 1.0 version
Any fixed can be always done in this or new change.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp (hunk 13 above)
Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp]: I think we can recursive to find the real file of non-header file. Tracking the include chain and find the first file which includes this decl.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp]: No need to do recursive, using (current) first declaration should also work.

File: clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp [hunk 15]
```diff
@@ -24,12 +24,28 @@ namespace {
...
+AST_MATCHER_P(Decl, isAllRedeclsInMainFile, FileExtensionsSet,
+              HeaderFileExtensions) {
   return llvm::all_of(Node.redecls(), [&](const Decl *D) {
-    SourceManager &SM = Finder->getASTContext().getSourceManager();
-    const SourceLocation L = D->getLocation();
-    return SM.isInMainFile(L) &&
-           !utils::isSpellingLocInHeaderFile(L, SM, HeaderFileExtensions);
+    return isInMainFile(D->getLocation(),
+                        Finder->getASTContext().getSourceManager(),
+                        HeaderFileExtensions);
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:48]: this isn't needed.
All you need to do is just get location and:
```
return SM.isInMainFile(L) ||  !utils::isSpellingLocInHeaderFile(L, SM, HeaderFileExtensions);
```

Or better add: isSpellingLocInSourceFile

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:48]: Still I would just assume that if first declaration isn't in header file, then such declaration could be considered to be internal. As include for header should be there anyway.

Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:48]: I prefer to tolerant some false negatives instead of introducing false positives for corner cases. It is more robust to check whether all redecls are in main file. And it will not cause performance issue since most of thing only have one declaration and one definition.

Comment (HerrCai0907) [reply clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp:48]: The reason of using recursive algorithm is there are some code practices will use macro and inc file to avoid duplicated code. In llvm-project, it is also used widely.
